File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
/******************************************************************************

                            Online Java Compiler.
                Code, Compile, Run and Debug java program online.
Write your code in this editor and press "Run" button to execute it.

*******************************************************************************/
import java.util.*; 
import java.io.*; 

public class Main
{
    static int[][] arr; 
    static int n,m,result; 
    static List<int[]> chiken; 
    static List<int[]> house; 
    
	public static void main(String[] args) throws IOException{
		//치킨집 조합 m 개 만들고 치킨거리가 가장 작은거 선택
		
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(bf.readLine()); 
		
		//배열 받아오기 
		n = Integer.parseInt(st.nextToken()); 
		m = Integer.parseInt(st.nextToken()); 
		arr = new int[n][n];
		house = new ArrayList<>();
		chiken = new ArrayList<>(); 
		result = 1000000000;
		
		for(int i=0; i<n; i++){
		    st = new StringTokenizer(bf.readLine()); 
		    for(int j=0; j<n; j++){
		        arr[i][j] = Integer.parseInt(st.nextToken()); 
		        if( arr[i][j] ==1){
		            house.add(new int[]{i,j});
		        }
		        if(arr[i][j] ==2){
		            chiken.add(new int[]{i,j});
		        }
		    }
		}
		
		//치킨집 조합 구하기 
		combi(0,0,new int[m][2]);
		
		
		System.out.println(result); 
	
	}
	
	static void combi(int start,int cnt,int[][]temp){
	    if(cnt==m){
	        //치킨거리 구하기 
	        sum(temp); 
	        return ;
	    }
	    
	    for(int i=start; i<chiken.size(); i++){
	        int[] now = chiken.get(i);
 	        temp[cnt][0] = now[0];
 	        temp[cnt][1] = now[1]; 
 	        combi(i+1,cnt+1,temp); 
	    }
	}
	
	static void sum(int[][] temp){
	    int answer = 0; 
	    
	    for(int[] h : house){
	        int x = h[0];
	        int y = h[1]; 
	        int houseSum = 1000000000;
	        for(int[] c : temp){
	            int sum = Math.abs(c[0]-x)+Math.abs(c[1]-y); 
	            houseSum = Math.min(houseSum,sum); 
	        }
	        answer+= houseSum;
	    }
	    result = Math.min(result,answer); 
	}
	
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
import sys
from collections import deque
input = sys.stdin.readline
#중복있게 n 중  m개를 고른 수열 -> 중복조합
n,m = map(int,input().split())
arr = [list(map(int,input().split())) for _ in range(n)]
#2를 m 개 고를때 1 과 2의 거리의 최솟값
chicken_arr = []
houses = []
answer = float('inf')
for i in range(n):
    for j in range(n):
        if arr[i][j] == 2:
            chicken_arr.append((i,j))
        if arr[i][j] ==1:
            houses.append((i,j))


def combination(cnt,start,minus_arr):
    if cnt == m:
        sum_chicken_dist(list(minus_arr)) #치킨집 조합
        return

    for i in range(start,len(chicken_arr)):
        temp = chicken_arr[i]
        minus_arr.append((temp[0],temp[1]))
        combination(cnt+1,i+1,minus_arr)
        minus_arr.pop()

def sum_chicken_dist(minus_arr):
    result = 0
    global answer
    for house in houses:
        dist = float('inf')
        for chicken_house in minus_arr: #조합별 집들과의 거리
            house_x = house[0]
            house_y = house[1]
            chicken_house_x = chicken_house[0]
            chicken_house_y = chicken_house[1]
            dist = min(dist,abs(house_x-chicken_house_x)+abs(house_y-chicken_house_y))
        result+=dist

    answer = min(answer,result)

combination(0,0,[])

print(answer)